templates/decays.hpp.in

```diff
@@ -0,0 +1,276 @@
+
+   void clear();
+   void clear_problems();
+   void calculate_decays();
+
+@decaysGetters@
+@decaysCalculationPrototypes@
+
+@partialWidthCalculationPrototypes@
+
+private:
+   @ModelName@_mass_eigenstates model{};
+   @ModelName@_input_parameters input{};
+   softsusy::QedQcd qedqcd{};
+   HigherOrderSMCorrections include_higher_order_corrections{HigherOrderSMCorrections::enable};
+   bool run_to_decay_particle_scale {true};
+   @ModelName@_decay_table decay_table{};
+   Decays_loop_corrections loop_corrections{};
+   Decays_problems problems{};
+
+   template<typename FieldIn, typename FieldOut1, typename FieldOut2>
+   typename Decay_amplitude_type<FieldIn, FieldOut1, FieldOut2>::type
+   calculate_amplitude(
+      const @ModelName@_cxx_diagrams::context_base&,
+      const typename @ModelName@_cxx_diagrams::field_indices<FieldIn>::type&,
+      const typename @ModelName@_cxx_diagrams::field_indices<FieldOut1>::type&,
+      const typename @ModelName@_cxx_diagrams::field_indices<FieldOut2>::type&) const;
+
+   template<typename FieldIn, typename FieldOut1, typename FieldOut2>
+   double amplitude_squared(@ModelName@_cxx_diagrams::context_base const& context,
+                  typename @ModelName@_cxx_diagrams::field_indices<FieldIn>::type const& indexIn,
+                  typename @ModelName@_cxx_diagrams::field_indices<FieldOut1>::type const& indexOut1,
+                  typename @ModelName@_cxx_diagrams::field_indices<FieldOut2>::type const& indexOut2) const;
+
+   template<typename FieldIn, typename FieldOut1, typename FieldOut2>
+   double get_partial_width(
+      const @ModelName@_cxx_diagrams::context_base&,
+      typename @ModelName@_cxx_diagrams::field_indices<FieldIn>::type const&,
+      typename @ModelName@_cxx_diagrams::field_indices<FieldOut1>::type const&,
+      typename @ModelName@_cxx_diagrams::field_indices<FieldOut2>::type const&) const;
+};
+
+@calcAmplitudeSpecializationDecls@
+
+template<typename FieldIn, typename FieldOut1, typename FieldOut2>
+double
+@ModelName@_decays::amplitude_squared(@ModelName@_cxx_diagrams::context_base const& context,
+                  typename @ModelName@_cxx_diagrams::field_indices<FieldIn>::type const& indexIn,
+                  typename @ModelName@_cxx_diagrams::field_indices<FieldOut1>::type const& indexOut1,
+                  typename @ModelName@_cxx_diagrams::field_indices<FieldOut2>::type const& indexOut2) const
+{
+
+   const auto mat_elem = calculate_amplitude<FieldIn, FieldOut1, FieldOut2>(
+      context, indexIn, indexOut1, indexOut2);
+   return mat_elem.square();
+}
+
+// returns a squared color generator for 3 point amplitude with FieldIn, FieldOut1 and FieldOut2
+// the generator is guessed from color representations of FieldIn, FieldOut1 and FieldOut2
+// This is not a bulletproof solution and might fail in general but is enough for
+// decays of color singlets
+template<typename FieldIn, typename FieldOut1, typename FieldOut2>
+double squared_color_generator() {
```

Comment:
It seems to me that this function is model independent and may be better put into `src/`.